conans/model/graph_lock.py

```diff
@@ -119,32 +128,44 @@ def __init__(self, graph=None):
         self._nodes = {}  # {numeric id: PREF or None}
         self.revisions_enabled = None
 
-        if graph:
+        if graph is not None:
             for node in graph.nodes:
                 if node.recipe == RECIPE_VIRTUAL:
                     continue
-                requires = {}
-                for edge in node.dependencies:
-                    requires[repr(edge.require.ref)] = edge.dst.id
-                # It is necessary to lock the transitive python-requires too, for this node
-                python_reqs = None
-                reqs = getattr(node.conanfile, "python_requires", {})
-                if isinstance(reqs, dict):  # Old python_requires
-                    python_reqs = {}
-                    while reqs:
-                        python_reqs.update(reqs)
-                        partial = {}
-                        for req in reqs.values():
-                            partial.update(getattr(req.conanfile, "python_requires", {}))
-                        reqs = partial
-
-                    python_reqs = [r.ref for _, r in python_reqs.items()]
-                elif isinstance(reqs, PyRequires):
-                    # If py_requires are defined, they overwrite old python_reqs
-                    python_reqs = node.conanfile.python_requires.all_refs()
-                graph_node = GraphLockNode(node.pref if node.ref else None, python_reqs,
-                                           node.conanfile.options.values, False, requires, node.path)
-                self._nodes[node.id] = graph_node
+                self._upsert_node(node)
+
+    def _upsert_node(self, node):
+        requires = []
+        build_requires = []
+        for edge in node.dependencies:
+            if edge.build_require:
+                build_requires.append(edge.dst.id)
+            else:
+                requires.append(edge.dst.id)
+        # It is necessary to lock the transitive python-requires too, for this node
+        python_reqs = None
+        reqs = getattr(node.conanfile, "python_requires", {})
+        if isinstance(reqs, dict):  # Old python_requires
+            python_reqs = {}
+            while reqs:
+                python_reqs.update(reqs)
+                partial = {}
+                for req in reqs.values():
+                    partial.update(getattr(req.conanfile, "python_requires", {}))
+                reqs = partial
+
+            python_reqs = [r.ref for _, r in python_reqs.items()]
+        elif isinstance(reqs, PyRequires):
+            # If py_requires are defined, they overwrite old python_reqs
+            python_reqs = node.conanfile.python_requires.all_refs()
+
+        previous = node.graph_lock_node
+        modified = previous.modified if previous else None
```

Comment:
The `previous` variable name is a bit misleading IMO, I was expecting it to be used afterward. I would just do 

```suggestion
        modified = node.graph_lock_node.modified if node.graph_lock_node else None
```